pkg/subctl/cmd/join.go

```diff
@@ -268,16 +270,28 @@ func joinSubmarinerCluster(config *rest.Config, subctlData *datafile.SubctlData)
 		err = checkOverlappingClusterCidr(globalNetworks)
 		status.End(err == nil)
 		exitOnError("Error validating overlapping ClusterCIDRs", err)
-	} else if globalNetworks[clusterID] == nil || globalNetworks[clusterID].GlobalCIDRs == nil || len(globalNetworks[clusterID].GlobalCIDRs) <= 0 {
-		// Globalnet enabled, no globalCidr configured on this cluster
-		globalCIDR, err = globalnet.AllocateGlobalCIDR(globalNetworks, subctlData)
-		status.End(err == nil)
-		status.QueueSuccessMessage(fmt.Sprintf("Allocated GlobalCIDR: %s", globalCIDR))
-		exitOnError("Globalnet failed", err)
+	} else if globalCIDR == "" {
+		// Globalnet enabled, GlobalCIDR not specified by the user
+		if cidrispreconfigured(clusterID, globalNetworks) {
+			// globalCidr already configured on this cluster
+			globalCIDR = globalNetworks[clusterID].GlobalCIDRs[0]
+			status.QueueSuccessMessage(fmt.Sprintf("Cluster already has GlobalCIDR allocated: %s", globalNetworks[clusterID].GlobalCIDRs[0]))
+		} else {
+			// no globalCidr configured on this cluster
+			globalCIDR, err = globalnet.AllocateGlobalCIDR(globalNetworks, subctlData)
+			status.End(err == nil)
+			status.QueueSuccessMessage(fmt.Sprintf("Allocated GlobalCIDR: %s", globalCIDR))
+			exitOnError("Globalnet failed", err)
+		}
 	} else {
-		// Globalnet enabled, globalCidr already configured on this cluster
-		globalCIDR = globalNetworks[clusterID].GlobalCIDRs[0]
-		status.QueueSuccessMessage(fmt.Sprintf("Cluster already has GlobalCIDR allocated: %s", globalNetworks[clusterID].GlobalCIDRs[0]))
+		// Globalnet enabled, globalCIDR specified by user
+		if cidrispreconfigured(clusterID, globalNetworks) {
+			// globalCidr pre-configured on this cluster
+			globalCIDR = globalNetworks[clusterID].GlobalCIDRs[0]
+			status.QueueFailureMessage(fmt.Sprintf("Pre-configured GlobalCIDR %s detected. Not changing it.", globalCIDR))
+		}
+		// globalCidr as specified by the user
+		status.QueueSuccessMessage(fmt.Sprintf("GlobalCIDR is: %s", globalCIDR))
```

Comment:
If user explicitly specifies the globalnet-cidr, we hit this case. Where are we validating that this globalnet-cidr does not match/overlap with any existing allocations?

@vthapar is experienced in this area of code. Vishal, can you please check if this is handled?

Reply:
This is missing following checks:
1. Global-cidr specified should be a valid CIDR, that should probably be done in some validation of arguments passed before we even discover networks.
2. Check if overlaps with existing global and throw error if it does.

Reply:
Check #2 could be prone to races.
In any case if the user specified his own CIDR I agree it is good to do input validation that it's a valid CIDR but I wouldn't bother with checking overlap. We don't want to babysit our users, if they specify one explicitly they can just as well make sure they are specifying a distinct one.

Reply:
you mean it does not match for 2 clusters?

Reply:
We have another issue to handle the race conditions by using a ConfigMap on the Broker cluster. So let's validate the overlap check.

Reply:
> If user explicitly specifies the globalnet-cidr, we hit this case. Where are we validating that this globalnet-cidr does not match/overlap with any existing allocations?
> 
you mean allocations for 2 clusters are same?
> @vthapar is experienced in this area of code. Vishal, can you please check if this is handled?

Reply:
> Check #2 could be prone to races.
> In any case if the user specified his own CIDR I agree it is good to do input validation that it's a valid CIDR but I wouldn't bother with checking overlap. We don't want to babysit our users, if they specify one explicitly they can just as well make sure they are specifying a distinct one.

Agree. It is upto the user to check for non-overlapping allocations

Reply:
> We have another issue to handle the race conditions by using a ConfigMap on the Broker cluster. So let's validate the overlap check.

Doesn't matter, the whole idea here is to let the user specify whatever CIDR he wants so lets keep it simple.
Adding this check is IMHO unnecessary complication which doesn't add real value.

Reply:
> you mean allocations for 2 clusters are same?

Currently when subctl auto-allocates globalnet-cidr from the --globalnet-cidr-range it ensures that its unique and does not overlap with existing allocations. Since this PR is adding support to specify --globalnet-cidr as part of join, before we consider the input as valid, subctl should validate the same checks on the user-provided input.

Reply:
> > We have another issue to handle the race conditions by using a ConfigMap on the Broker cluster. So let's validate the overlap check.
> 
> Doesn't matter, the whole idea here is to let the user specify whatever CIDR he wants so lets keep it simple.
> Adding this check is IMHO unnecessary complication which doesn't add real value.

Noooo, it's not complicated. We already have APIs to check if there is overlap in the CIDRs. I don't understand why you are saying that lets blindly trust the user input when we can easily validate it (and save time in debugging failures that happen at run-time).

Reply:
> > > We have another issue to handle the race conditions by using a ConfigMap on the Broker cluster. So let's validate the overlap check.
> > 
> > 
> > Doesn't matter, the whole idea here is to let the user specify whatever CIDR he wants so lets keep it simple.
> > Adding this check is IMHO unnecessary complication which doesn't add real value.
> 
> Noooo, it's not complicated. We already have APIs to check if there is overlap in the CIDRs. I don't understand why you are saying that lets blindly trust the user input when we can easily validate it (and save time in debugging failures that happen at run-time).

Trust me, debugging any issues which are a result of overlapping globalCIDRs will be extremely hard. This is a simple validation IMO @mkolesnik

Reply:
> > you mean allocations for 2 clusters are same?
> 
> Currently when subctl auto-allocates globalnet-cidr from the --globalnet-cidr-range it ensures that its unique and does not overlap with existing allocations. Since this PR is adding support to specify --globalnet-cidr as part of join, before we consider the input as valid, subctl should validate the same checks on the user-provided input.

This is expected though since its part of the basics of the operating mechanism of the way it allocates CIDRs.

> Noooo, it's not complicated. We already have APIs to check if there is overlap in the CIDRs. I don't understand why you are saying that lets blindly trust the user input when we can easily validate it (and save time in debugging failures that happen at run-time).

Why do you think you're smarter than the user in this case?
Perhaps the user wants to do it for a certain reason, it has nothing to do with validation being easy or hard.

Reply:
> Why do you think you're smarter than the user in this case?

lol, this is not the point of discussion.

> Perhaps the user wants to do it for a certain reason, it has nothing to do with validation being easy or hard.

User can provide any input (sometimes intentional and sometimes unintentional) and we have checks in code to ensure that we gracefully handle wrong input and warn him at a very early stage. The question is - does Submariner Globalnet support "Overlapping" globalCIDRs. The plain answer is NO, so we should warn him at the deployment stage itself.

Reply:
> User can provide any input (sometimes intentional and sometimes unintentional) and we have checks in code to ensure that we gracefully handle wrong input and warn him at a very early stage. The question is - does Submariner Globalnet support "Overlapping" globalCIDRs. The plain answer is NO, so we should warn him at the deployment stage itself.

OK than I propose a settlement - that we can do a "detection" if you desire, but not block the operation. Print a warning so the user knows, so he can know and decide if that's what he wanted or not.

Reply:
I agree with sridhar here, we must do validation. If we have a way of detecting how the admin is shooting his feet, lets prevent it. Think of this as a gun safelock... 

What we couldn't prevent for example is the admin editing the ConfigMap on the broker (when we have that..) and that will be his way to override things if the admin knows what he's doing

Reply:
If an admin forces the use of a conflicting CIDR, what would the consequences be? If it could break a running cluster, then it seems to me we should prevent it by default...

Reply:
Checking user input for invalid inputs is UX 101. Adding a cluster with overlapping CIDR can break things. Not letting user shoot itself in the foot is always good design. Not doing validations coz there is a race condition sounds more like a race condition problem that should be fix than validation problem.

Reply:
By your guys standards then `rm` should prevent running on `/` if `-r` is specified, since the user is stupid and can seriously screw up something.
Yet, we have many systems and languages that allow users more flexibility and dont babysit him, even if he can "shoot himself in the foot".

So if you want to be holier than the pope, by all means cotton wrap the user so the poor thing doesnt break anything by accident. I prefer to walk in the footsteps of giants and let him shoot himself if the input is valid :)

* And yes, even windows which is much more restrictive to a "dumb user" will allow him to format his OS drive quite easily..

Reply:
The consequences I believe would be conflicting iptable rules, and route rules, may be the route agent would end in a crash loop (not sure of how error handling happens there)... @sridhargaddam probably you have more insight.

Something that came up while talking with mike about this, is that... at a minimun we may need to support removal of globalCIDRs from clusters, ... 

For example, if you keep adding and removing clusters, ... we don't handle cleanups properly... or what if you gave more CIDRs to a cluster than you wanted?, there must be a way to fix it.

Reply:
> The consequences I believe would be conflicting iptable rules, and route rules, may be the route agent would end in a crash loop (not sure of how error handling happens there)... @sridhargaddam probably you have more insight.

If the user-provides a globalnet-cidr that overlaps with globalnet-cidr of remote cluster, then cross-cluster connectivity will fail. I dont think it will affect local cluster traffic. 
Regarding a crash in route-agent, I don't think so based on my understanding of route-agent code. When overlapping globalCIDRs are used during deployment, it would make debugging lot more difficult as there will not be any crash but traffic could go in totally different ways than what we anticipate/program. So, it's better to ensure that we are validating the input at deployment time.

> 
> Something that came up while talking with mike about this, is that... at a minimun we may need to support removal of globalCIDRs from clusters, ...

IMHO, I would say that we should start looking at cleanup (of submariner resources) in general which is missing today. 

> 
> For example, if you keep adding and removing clusters, ... we don't handle cleanups properly... or what if you gave more CIDRs to a cluster than you wanted?, there must be a way to fix it.

Right, this is more about updating the globalCIDR of a cluster where updation can be expanding/shrinking/allocating-newCIDR. Yes, we should support this use-case as a new PR.

Reply:
+1 to not giving the user the option to shoot themselves.

Reply:
> By your guys standards then `rm` should prevent running on `/` if `-r` is specified, since the user is stupid and can seriously screw up something.

`rm` (at least, GNU `rm`) does check for that ;-).